File: ThreadPool/waitgroup.swift
import Foundation

public final class WaitGroup {

    var index: Int

    let mutex: Mutex

    let condition: Condition

    public init() {
        index = 0
        mutex = Mutex()
        condition = Condition()
    }

    public func enter() {
        mutex.whileLocked {
            index += 1
        }
    }

    public func done() {
        mutex.whileLocked {
            guard index >= 1 else { return }
            index -= 1
            if index == 0 {
                condition.broadcast()
            }
        }
    }

    public func waitForAll() {
        mutex.whileLocked {
            condition.wait(mutex: mutex, condition: index == 0)
        }
    }
}

File: CXX_Thread/waitgroup.cpp
#include "waitgroup.h"

void waitgroup::enter()
{
    auto lock = unique_lock{this->mtx};
    this->count++;
}

void waitgroup::done()
{
    auto lock = unique_lock{this->mtx};
    if (this->count < 1)
        return;
    this->count--;
    if (this->count == 0)
        this->cv.notify_all();
}

void waitgroup::wait()
{
    auto lock = unique_lock{this->mtx};
    this->cv.wait(lock, [&]()
                  { return this->count == 0; });
}

waitgroup *_Nonnull waitgroup::init()
{
    return new waitgroup();
}